### PR TITLE
[RESTEASY-2925]: Inconsitent NewCookie parsing when unknown attributes is found

### PR DESCRIPTION
**Pull request by NicoNes**
https://issues.redhat.com/browse/RESTEASY-2925

### CODE REVIEW

File: resteasy-core/src/main/java/org/jboss/resteasy/plugins/delegates/NewCookieHeaderDelegate.java [hunk 1]
```diff
@@ -35,12 +37,16 @@ public NewCookie fromString(String newCookie) throws IllegalArgumentException {
       Date expiry = null;
 
       ParameterParser parser = new ParameterParser();
-      Map<String, String> map = parser.parse(newCookie, ';');
+      LinkedHashMap<String, String> map = parser.parse(newCookie, ';');
```
Comment (jamezp) [inline resteasy-core/src/main/java/org/jboss/resteasy/plugins/delegates/NewCookieHeaderDelegate.java:40]: There is not need for this to be a `LinkedHashMap`. This line can remain unchanged.

File: resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java [hunk 2]
```diff
@@ -223,11 +222,11 @@ public void setLowerCaseNames(boolean b)
     * @param separator the name/value pairs separator
     * @return a map of name/value pairs
     */
-   public Map<String, String> parse(final String str, char separator)
+   public LinkedHashMap<String, String> parse(final String str, char separator)
```
Comment (jamezp) [inline resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java:225]: We shouldn't change the return type here.

File: resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java [hunk 3]
```diff
@@ -241,11 +240,11 @@ public void setLowerCaseNames(boolean b)
     * @param separator the name/value pairs separator
     * @return a map of name/value pairs
     */
-   public Map<String, String> parse(final char[] chars, char separator)
+   public LinkedHashMap<String, String> parse(final char[] chars, char separator)
```
Comment (jamezp) [inline resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java:243]: We shouldn't change the return type here either.

File: resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java [hunk 4]
```diff
@@ -261,7 +260,7 @@ public void setLowerCaseNames(boolean b)
     * @param separator the name/value pairs separator
     * @return a map of name/value pairs
     */
-   public Map<String, String> parse(
+   public LinkedHashMap<String, String> parse(
```
Comment (jamezp) [inline resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java:263]: We shouldn't change the return type here either.

File: resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java [hunk 5]
```diff
@@ -270,9 +269,9 @@ public void setLowerCaseNames(boolean b)
 
       if (chars == null)
       {
-         return new HashMap<String, String>();
+         return new LinkedHashMap<>();
       }
-      HashMap<String, String> params = new HashMap<String, String>();
+      LinkedHashMap<String, String> params = new LinkedHashMap<>();
```
Comment (jamezp) [inline resteasy-core/src/main/java/org/jboss/resteasy/util/ParameterParser.java:274]: This should just be `Map<String, String> params = new LinkedHashMap<>()`.

**Comment by NicoNes** [comment]
In this PR I changed the `HashMap<String,String>` used in the parser by a `LinkedHashMap<String,String>`.

I could have let the `parse(...)` methods signatures unchanged but I decided to change them to replace the `Map<String,String>` interface by the `LinkedHashMap<String,String>`  to make it clear that order matter.

Let me know what you think.

**Review by jamezp — CHANGES_REQUESTED** [review]

**Comment by jamezp** [comment]
> In this PR I changed the `HashMap<String,String>` used in the parser by a `LinkedHashMap<String,String>`.
> 
> I could have let the `parse(...)` methods signatures unchanged but I decided to change them to replace the `Map<String,String>` interface by the `LinkedHashMap<String,String>` to make it clear that order matter.
> 
> Let me know what you think.

I don't think we should change the signatures. That can break backwards compatibility.

**Comment by NicoNes** [comment]
@jamezp Even with a `LinkedHashMap` in the signature, backwards compatibility is still OK since a `LinkedHashMap`  is a `Map` and also a `HashMap` right ?
Actually I did the change because it can be be weird to say that cookie name is always the first attribute and then retrieve it iterating over a `Map` interface that does not enforce order.
What bother me more with using `LinkedHasMap` is that it is an implementation not an interface...
I will change it.

Thanks

**Review by jamezp — APPROVED** [review]
Looks good thanks @NicoNes!

**Pull request merged**